network.go

```diff
@@ -104,19 +106,33 @@ func (n *Network) SetMetricsHook(f func(pm map[string]PeerMetrics)) {
 }
 
 // Run starts the network.
-// Listens to incoming connections on the specified port
-// and connects to other peers
-func (n *Network) Run() {
-	n.logger.Infof("Starting a P2P Network with configuration %+v", n.conf)
+// Listens to incoming connections on the specified port and connects to other peers
+func (n *Network) Run() error {
+	select {
+	case <-n.stopper:
+		return fmt.Errorf("unable to restart a network that has been stopped")
+	default:
+		n.logger.Infof("Starting a P2P Network with configuration %+v", n.conf)
+		n.controller.Start()
+		return nil
+	}
 
-	n.controller.Start() // this will get peer manager ready to handle incoming connections
-	//DebugServer(n)
 }
 
-func (n *Network) Stop() {
-	// TODO implement
-	// close stop channel
-	// stop all peers
+// Stop shuts down the network
+// Note that the network object will become unusable after it is stopped
+func (n *Network) Stop() error {
+	select {
+	case <-n.stopper:
+		return fmt.Errorf("network not running")
```

Comment:
"network already stopped" is slightly better wording? Because your wording would include the case where the network has not been started yet, but in this case it wouldn't go in this case (because stopper would still be open but empty).